include/fmt/format-inl.h

```diff
@@ -272,14 +302,17 @@ inline uint64_t multiply(uint64_t lhs, uint64_t rhs) {
 #endif
 }
 
-inline fp operator*(fp x, fp y) { return {multiply(x.f, y.f), x.e + y.e + 64}; }
+FMT_CONSTEXPR inline fp operator*(fp x, fp y) {
+  return {multiply(x.f, y.f), x.e + y.e + 64};
+}
 
 // Returns a cached power of 10 `c_k = c_k.f * pow(2, c_k.e)` such that its
 // (binary) exponent satisfies `min_exponent <= c_k.e <= min_exponent + 28`.
-inline fp get_cached_power(int min_exponent, int& pow10_exponent) {
+FMT_CONSTEXPR inline fp get_cached_power(int min_exponent,
+                                         int& pow10_exponent) {
   // Normalized 64-bit significands of pow(10, k), for k = -348, -340, ..., 340.
   // These are generated by support/compute-powers.py.
-  static constexpr const uint64_t pow10_significands[] = {
```

Comment:
So you ran into the problem of `static` function block scope variables (with initialization on first use a.k.a 'magic statics') being invalid in `constexpr` context. Giving them automatic duration means initialization on every call.

I recommend moving `pow10_significands[]` and `pow10_exponents` into `basic_data` similar to what I have been proposing in #2352

Reply:
Actually, I thought that compilers are able to optimize this code (based on the fact that this approach is already used in the library - `power_of_10_64` or `bsr2log10`), but I was wrong making this assumption: https://godbolt.org/z/KWYhcqKxn (Clang ❤️).

Quick bench shows that my change made `get_cached_power` _12 times_ slower on GCC, such a great result:

- GCC bench - https://quick-bench.com/q/2p1fSVAl01RjGGSD8a0lan5OPfU
- Clang bench - https://quick-bench.com/q/VB_AUt8Eeb_QivHXd0C1HB5zA2g

So, yeah, I should definitely fix this.

Reply:
Yeah, it's used elsewhere and both gcc and msvc are still suffering from the problem that I've been pointing out a month ago. I think #2352 should be reopened and reworked.

Reply:
@DanielaE, about your recommendation - should `pow10_significands` and `pow10_exponents` here (and maybe some other variables in other functions) be moved to `basic_data` only, or maybe they can be just moved outside from the function as separate `static constexpr` variables?

Reply:
It's about possible duplication of tables in different TUs. If you make them members of `basic_data` then you get guaranteed deduplication across the board because of the age of templates and maturity of implementations. With `static constexpr` you may end up with as many copies as there are TUs #including the respective headers. `constexpr` variables by themselves are language entities with _external_ linkage (i.e. generation of linker symbols). Adding `static` changes linkage to _internal_. It's then up to the compiler and sheer happenstance to deduplicate the tables in the final program. A totally valid implementation is neither required nor encouraged to analyse the TUs further to attempt deduplication because there are so many opportunities to trigger ODR-use and implementations aren't required to check for the possibility of ODR-use from a non-`constexpr` context to _guarantee_ no runtime access to these table.

Reply:
Ok, thank you for such comprehensive description.
I'm just a bit disappointed that we have to move these initializers back to `basic_data` in `format.h`. Maybe we can reuse the same idea for `format-inl.h` since neither `pow10_significands` nor `pow10_exponents` is used in `format.h`.

Reply:
What's the reason for disappointment? Is it the need to compile the tables? With the definition of these tables in `format.h` in full view, compilers can take advantage of that. Shifting the definition to the point of instantiation in `format-inl.h` may buy you some microseconds of compilation speed but hides the actual values from the compiler. It's a matter of priorities.

Reply:
No, I'm disappointed that they were the fields of `basic_data` in `format.h`, if I'm correct, later they were moved here, and now we are about to move it back. And since these variables are not needed in `format.h`, then we can probably create some similar `basic_data` that is specific for `format-inl.h`.
So it's not about some compilation-time improvements (also, I'm pretty sure that `constexpr` variables should be defined anyway, it's another big *no* for this kind of optimization), but about scoping variables to only where they are needed.

Reply:
They moved to the newly created `basic_inl_data` templated class. I also moved `power_of_10_64` there, since codegen differs between `static`/`constexpr` variable in the function.